Plugins for chat

sigorilla opened the pull request:
Сделал серверную часть для модулей. 

Модули создаются как обычные NPM пакеты. Главное чтобы они имели префикс `shriek-`. А так же переменную для экспорта `forEvent`. 

Названия событий для `forEvent` будут позже.

**Примеры**
- [Markdown](https://github.com/sigorilla/shriek-markdown)
- [Emoji](https://github.com/sigorilla/shriek-emoji)
- [Open Graph](https://github.com/sigorilla/shriek-opengraph)

sigorilla committed:
markdown module for backend

sigorilla committed:
upd readme [skip ci]

sigorilla committed:
remove module showdown

sigorilla committed:
add emoji package and change require of module

sigorilla commented:
Для markdown важен порядок (он должен быть раньше всех), поэтому пока я сделал обычный список из подключения, а не автоматический.

sigorilla committed:
update emoji for react

sigorilla committed:
update emoji package

AndrewUldin commented on app/modules/channel.js:
```diff
@@ -136,6 +138,12 @@ var channelModule = function (socket) {
         if (!err) {
+          shriekModules.forEach(function (module) {
+            if (module.forEvent === 'channelGet') {
+              data = module(data);
```
вот тут надо обернуть в промис, либо для каждого, либо весь forEach

sigorilla replied:
А если выше уже обернуто в промис, то надо это снова делать?

AndrewUldin commented on app/modules/message.js:
```diff
@@ -25,6 +27,12 @@ var messageModule = function(socket) {
       if (!err) {
+        shriekModules.forEach(function (module) {
+          if (module.forEvent === 'channelGet') {
+            data[0] = module([data]);
```
тут тоже обернуть в промис

sigorilla committed:
add promise

sigorilla committed:
add Promise to save message

sigorilla committed:
Resolve conflicts

sigorilla committed:
resolve conflicts

AndrewUldin commented:
:+1:

sigorilla committed:
change logic of modules with callback

sigorilla committed:
update opengraph and upd readme [skip ci]

sigorilla committed:
move styles and images to sigorilla/shriek-emoji

Fix #83

sigorilla committed:
Resolve conflicts with develop

sigorilla commented:
@shri-2015-org/shriek 
Возможно надо, чтобы модули отрабатывали только при отправке сообщения. Тем самым мы могли бы сначала показывать обычное сообщение. А после того, как оно сохранилось, мы присылаем обновленный вариант и заменяем на новый.
И при этом не надо каждый раз нагружать сервер обработкой всех сообщений в массиве при получении канала.

sigorilla committed:
Update logic of modules

Теперь модули изменяют текст сообщения до сохранения. Тем самым в `channel get` не надо вызывать модули на каждое сообщение при получении.

sigorilla committed:
change logic for react components

sigorilla committed:
rename modules to plugins

sigorilla commented:
@verkholantsev Попробовал компонент реакты подключить по-другому. Теперь записываются в глобальный массив, который потом иницилизирует в экшенах сообщений (или если захотеть, то и других компонентах). 

Был вариант сделать экшен глобальный для регистрации плагина, но бовер подключает скрипты раньше, чем наш основной файл. Следовательно этой функции в глобале не будет. Ну и соответственно ошибки. 
Если @shri-2015-org/shriek подскажет, как бовером (точнее wiredep) можно разные файлы в разные места в `index.html` вставить, тогда можно будет функцию регистрации сделать.

P.S. После этого уже зарезовлю конфликты.

sigorilla committed:
remove global array, add registration function

sigorilla committed:
update readme

sigorilla commented:
Собственно отказался от глобального массива и смог распределить нужные файлы в необходимое место `index.html`. Всем **wiredep** и регулярки! Теперь плагин можно регистрировать в _системе_ плагинов.

sigorilla committed:
add raw field

sigorilla committed:
if field 'raw' empty show field 'text'

sigorilla committed:
merge with develop

sigorilla committed:
another merge with develop
